sbt-version-policy/src/main/scala/sbtversionpolicy/DependencyCheckReport.scala

```diff
@@ -61,44 +62,120 @@ object DependencyCheckReport {
-        case None => Missing(ver)
-        case Some(`ver`) => SameVersion(ver)
+      val status = currentModules.get(orgAndName) match {
+        case None => Missing(previousVersion)
+        case Some(`previousVersion`) => SameVersion(previousVersion)
         case Some(currentVersion) =>
           val reconciliation = reconciliations
             .collectFirst {
               case (matcher, rec) if matcher.matches(org, name) =>
                 rec
             }
             .getOrElse(defaultReconciliation)
-          if (reconciliation.isCompatible(ver, currentVersion))
-            CompatibleVersion(currentVersion, ver, reconciliation)
+          val isCompatible =
+            if (compatibilityIntention == Compatibility.BinaryAndSourceCompatible) {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation) &&
+                isSourceCompatible(currentVersion, previousVersion, reconciliation)
+            } else {
+              isBinaryCompatible(currentVersion, previousVersion, reconciliation)
+            }
+          if (isCompatible)
+            CompatibleVersion(currentVersion, previousVersion, reconciliation)
           else
-            IncompatibleVersion(currentVersion, ver, reconciliation)
+            IncompatibleVersion(currentVersion, previousVersion, reconciliation)
       }
 
-      orgName -> status
+      orgAndName -> status
+    }
+
+  private def isBinaryCompatible(currentVersion: String, previousVersion: String, versionCompatibility: VersionCompatibility): Boolean =
+    versionCompatibility.isCompatible(previousVersion, currentVersion)
+
+  private def isSourceCompatible(currentVersion: String, previousVersion: String, versionCompatibility: VersionCompatibility): Boolean =
+    versionCompatibility match {
+      case VersionCompatibility.Always =>
+        true
+      case VersionCompatibility.Strict | VersionCompatibility.Default | VersionCompatibility.PackVer =>
+        // In PVP, any release can break source compatibility
+        previousVersion == currentVersion
+      case VersionCompatibility.SemVer | VersionCompatibility.EarlySemVer =>
+        val (currentMajor, currentMinor, currentPatch) = extractSemVerNumbers(currentVersion)
+        val (previousMajor, previousMinor, previousPatch) = extractSemVerNumbers(previousVersion)
+        currentMajor == previousMajor && {
+          if (currentMajor == 0)
+            currentMinor == previousMinor && currentPatch == previousPatch
+          else
+            currentMinor == previousMinor && currentPatch >= previousPatch
+        }
+      case VersionCompatibility.SemVerSpec =>
+        val (currentMajor, currentMinor, currentPatch) = extractSemVerNumbers(currentVersion)
+        val (previousMajor, previousMinor, previousPatch) = extractSemVerNumbers(previousVersion)
+        currentMajor == previousMajor &&
+          currentMinor == previousMinor &&
+          currentPatch >= previousPatch
     }
```

Comment:
We implement source compatibility checking here, according to the versioning scheme of the dependency.